metric/metric.go

```diff
@@ -2,26 +2,161 @@ package metric
+			Name:      "results_http_ssl_last_chain_expiry_timestamp_seconds",
+			Help:      "Returns last SSL chain expiry in timestamp seconds",
+		}, []string{"key", "group", "name", "url"})
+	}
+}
+
 // PublishMetricsForEndpoint publishes metrics for the given endpoint and its result.
 // These metrics will be exposed at /metrics if the metrics are enabled
 func PublishMetricsForEndpoint(endpoint *core.Endpoint, result *core.Result) {
-	if resultCount == nil {
-		resultCount = promauto.NewCounterVec(prometheus.CounterOpts{
-			Name: "gatus_results_total",
-			Help: "Number of results per endpoint",
-		}, []string{"key", "group", "name", "success"})
+	ensurePrometheusMetrics()
+
+	resultCount.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL, strconv.FormatBool(result.Success)).Inc()
+	resultSuccessGauge.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL).Set(map[bool]float64{true: 1, false: 0}[result.Success])
+	resultDurationGauge.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL).Set(float64(result.Duration.Seconds()))
+
+	switch {
+	// DNS
+	case endpoint.DNS != nil:
+		resultDNSReturnCode.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL, result.DNSRCode).Inc()
+	// TCP
+	case strings.HasPrefix(endpoint.URL, "tcp://"):
+		if result.Connected {
+			resultTCPConnected.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL).Inc()
+		}
+	// ICMP
+	case strings.HasPrefix(endpoint.URL, "icmp://"):
+	// STARTTLS
+	case strings.HasPrefix(endpoint.URL, "starttls://"):
+		if result.Connected {
+			resultStartTLSConnected.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL).Inc()
+		}
+		if result.CertificateExpiration != 0 {
+			resultStartTLSSSLLastChainExpiryTimestampSeconds.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL).Set(float64(result.CertificateExpiration / time.Second))
+		}
+	// TLS
+	case strings.HasPrefix(endpoint.URL, "tls://"):
+		if result.Connected {
+			resultTLSConnected.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL).Inc()
+		}
+		if result.CertificateExpiration != 0 {
+			resultTLSSSLLastChainExpiryTimestampSeconds.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL).Set(float64(result.CertificateExpiration / time.Second))
+		}
+	// HTTP
+	default:
+		if result.Connected {
+			resultHTTPConnected.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL).Inc()
+		}
+		if result.HTTPStatus != 0 {
+			resultHTTPStatusCode.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL, strconv.Itoa(result.HTTPStatus)).Inc()
+		}
+		if result.CertificateExpiration != 0 {
+			resultHTTPSSLLastChainExpiryTimestampSeconds.WithLabelValues(endpoint.Key(), endpoint.Group, endpoint.Name, endpoint.URL).Set(float64(result.CertificateExpiration / time.Second))
+		}
 	}
```

Comment:
A lot of these look very similar to the code that can be found at https://github.com/TwiN/gatus/blob/a9fc876173a9cb50507d12c1a45a610af99444ae/core/endpoint.go#L232

If possible, I'd like to avoid duplication. Perhaps you could add a method such as `Type` in the `Endpoint` struct and have that logic only there instead of at two different places?

Reply:
add a method in core.Endpoint.Type() for get endpoint type.